packages/serve/test/auth/cannerPATAuthenticator.spec.ts

```diff
@@ -0,0 +1,277 @@
+import * as sinon from 'ts-sinon';
+import axios from 'axios';
+import { IncomingHttpHeaders } from 'http';
+import { Request, BaseResponse } from 'koa';
+import { CannerPATAuthenticator } from '@vulcan-sql/serve/auth';
+import { AuthResult, AuthStatus, KoaContext } from '@vulcan-sql/serve/models';
+
+jest.mock('axios');
+const mockedAxios = axios as jest.Mocked<typeof axios>;
+
+const getStubAuthenticator = async (
+  options: any,
+  stubValue: any = null
+): Promise<CannerPATAuthenticator> => {
+  const authenticator = new CannerPATAuthenticator({ options }, '');
+  if (stubValue) {
+    // TODO: refactor to avoid stubbing private method
+    // stub the private method fetchCannerUser to simulate the response from remote server
+    sinon.default
+      .stub(authenticator, <any>'fetchCannerUser')
+      .resolves(stubValue);
+  }
+  await authenticator.activate();
+  return authenticator;
+};
+
+const expectIncorrect = {
+  status: AuthStatus.INDETERMINATE,
+  type: 'canner-pat',
+};
+const expectFailed = (
+  message = 'authenticate user by "canner-pat" type failed.'
+) => ({
+  status: AuthStatus.FAIL,
+  type: 'canner-pat',
+  message,
+});
+const mockOptions = {
+  'canner-pat': {
+    host: 'mockHost',
+    port: 3000,
+    ssl: false,
+  },
+};
+const invalidToken = Buffer.from('clientId:clientSecret').toString('base64');
+
+it.each([
+  { 'canner-pat': { host: 'mockHost' } },
+  { 'canner-pat': { port: 3000 } },
+  { 'canner-pat': { ssl: false } },
+  { 'canner-pat': {} },
+])('Should throw configuration error when options = %p', async (options) => {
+  // Arrange
+  const ctx = {
+    ...sinon.stubInterface<KoaContext>(),
+    request: {
+      ...sinon.stubInterface<Request>(),
+      headers: {
+        authorization: 'Canner-PAT 1234567890',
+      },
+    },
+  } as KoaContext;
+  const authenticator = await getStubAuthenticator(options);
+
+  // Act, Assert
+  await expect(authenticator.authCredential(ctx)).rejects.toThrow(
+    'please provide correct connection information to Canner Enterprise, including "host" and "port".'
+  );
+});
+
+it('Should throw error when catch error when fetching the remote server for user info', async () => {
+  // Arrange
+  const ctx = {
+    ...sinon.stubInterface<KoaContext>(),
+    request: {
+      ...sinon.stubInterface<Request>(),
+      headers: {
+        ...sinon.stubInterface<IncomingHttpHeaders>(),
+        authorization: `Canner-PAT eyValidToken`,
+      },
+    },
+  } as KoaContext;
+  // mock axios.post method get non 200 response from remote server then axios throw error
+  mockedAxios.post.mockRejectedValue({
+    request: 'mock request',
+    response: {
+      data: { error: 'mock response error' },
+      status: 500,
+      statusText: 'string',
+      headers: {},
+      config: {},
+    } as any,
+    isAxiosError: true,
+    toJSON: () => ({}),
+    message: 'An error occurred!',
+    name: 'Error',
+  });
```

Comment:
Why do you need to mock the `axios.post` reject value ? You could mock the rejected value for `fetchCannerUser` directly.

Reply:
After discussion, in actually, @onlyjackfrost would like to test the `fetchCannerUser`'s catch and transform part, but currently `fetchCannerUser` is a private method that is called by the private `validate` method, so that why he needs to mock the `axio.post`, so our conclusion is to update the description fo the test case to make others know the intention.